File: design-patterns/java/State-DesignPattern/src/WithStatePattern/HasCardState.java
package WithStatePattern;

public class HasCardState implements ATMState {

    public void insertCard(ATMMachine atm) {
        System.out.println("Card already inserted");
    }

    public void enterPin(ATMMachine atm, int pin) {

        if(pin == 1234){
            System.out.println("PIN correct");
            atm.setState(new AuthenticatedState());
        } else {
            System.out.println("Wrong PIN");
            atm.setState(new IdleState());
        }
    }

    public void withdrawCash(ATMMachine atm, int amount) {
        System.out.println("Enter PIN first");
    }

    public void ejectCard(ATMMachine atm) {
        System.out.println("Card ejected");
        atm.setState(new IdleState());
    }
}

File: design-patterns/python/State-DesignPattern/src/with_state_pattern/has_card_state.py
"""HasCard state - card inserted, PIN not yet entered."""
from .atm_state import ATMState
from .atm_machine import ATMMachine


class HasCardState(ATMState):
    def insert_card(self, atm: ATMMachine) -> None:
        print("Card already inserted")

    def enter_pin(self, atm: ATMMachine, pin: int) -> None:
        from .idle_state import IdleState
        from .authenticated_state import AuthenticatedState
        if pin == 1234:
            print("PIN correct")
            atm.set_state(AuthenticatedState())
        else:
            print("Wrong PIN")
            atm.set_state(IdleState())

    def withdraw_cash(self, atm: ATMMachine, amount: int) -> None:
        print("Enter PIN first")

    def eject_card(self, atm: ATMMachine) -> None:
        from .idle_state import IdleState
        print("Card ejected")
        atm.set_state(IdleState())
